### PR TITLE
Implementation of Database Migration UI

### PR DESCRIPTION
**Pull request by tgqiang**
**Important files to look at:**
* index.py
* database_migrate.py
* databaseMigrate.html
* index.html

**How the page works:**

image

When the AY that is detected from the current date & time does not match the referential AY used by the system, a notice will appear in the home page. There will be a link in that notice, which allows users to direct themselves to the database-migration page.

image

There will be a checkbox and a button at the end of the disclaimer. Checking the checkbox will reveal the 'Proceed' button.

Please read through the disclaimer to check if the disclaimer is sufficiently clear.

Note: This fixes issue #77, #78

### CODE REVIEW

**Comment by xaterz** [comment]
Please add a check to /migrateDatabase for whether the new AY has arrived. If it hasn't, should redirect user back to home page.

**Comment by helloqx** [comment]
You are merging into master directly

File: components/handlers/index.py [hunk 1]
```diff
@@ -20,4 +20,15 @@ def GET(self):
         if not session.validate_session():
             raise web.seeother('/login')
         else:
-            return RENDER.index()
+            current_ay_with_current_date = model.get_current_ay_sem()[0:8]
+            current_database_ay = model.get_all_fixed_ay_sems()[0][0][0:8]
```
Comment (xaterz) [inline components/handlers/index.py]: Line 24: This can simply be model.get_current_ay()

File: components/handlers/database_migrate.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    '''
+    def GET(self):
+        '''
+            This function is called when the database migration page is loaded
+            If user is not logged in, they are redirected to the login page.
+        '''
+        if not session.validate_session():
+            raise web.seeother('/login')
+        else:
+            current_database_ay = model.get_all_fixed_ay_sems()[0][0][0:8]
```
Comment (xaterz) [inline components/handlers/database_migrate.py]: Line 23: Can simply use model.get_current_ay()

File: components/database.py [hunk 3]
```diff
@@ -1256,3 +1455,160 @@ def clean_old_sessions(date_to_delete):
...
+
+
+def update_active_modules_after_migration():
+    '''
+        This function makes sure that all modules in fixed mounting
+        have their statuses updated to "Active" from "New"
+    '''
+    sql_command = "UPDATE module SET status='Active' WHERE " + \
+                  "status='New' AND code in (SELECT mm.moduleCode FROM " + \
+                  "moduleMounted mm WHERE mm.moduleCode = code)"
```
Comment (xaterz) [inline components/database.py]: Line 1579: I think the 'WHERE mm.moduleCode = code)' is not necessary?

File: components/database.py [hunk 4]
```diff
@@ -1256,3 +1455,160 @@ def clean_old_sessions(date_to_delete):
...
+        6) Increment all students' year of study by 1.
+    '''
+    migrate_fixed_to_past_mounting()
+    migrate_one_ay_from_tentative_to_fixed()
+    duplicate_data_into_tentative_if_needed()
+    clean_up_module_backup_table()
+    update_active_modules_after_migration()
+    increment_student_year_by_one()
+
+    CONNECTION.commit()
```
Comment (xaterz) [inline components/database.py]: Line 1485: Need to wrap CONNECTION.commit() with try except statement

File: components/handlers/database_migrate.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            next_ay = model.get_next_ay(current_database_ay)
+            future_ay = model.get_next_ay(next_ay)
+
+            return RENDER.databaseMigrate(current_database_ay, next_ay, future_ay)
+
+
+    def POST(self):
+        model.migrate_to_next_aysem()
```
Comment (xaterz) [inline components/handlers/database_migrate.py]: Need to show outcome message for whether migration is successful or not.
migrate_to_next_aysem() should return a boolean to indicate the success/failure

File: components/handlers/index.py [hunk 6]
```diff
@@ -20,4 +20,15 @@ def GET(self):
         if not session.validate_session():
             raise web.seeother('/login')
         else:
-            return RENDER.index()
+            current_ay_with_current_date = model.get_current_ay_sem()[0:8]
+            current_database_ay = model.get_current_ay()
+
+            to_migrate_db = False
+
+            if current_ay_with_current_date != current_database_ay:
```
Comment (helloqx) [inline components/handlers/index.py]: Possible to use < instead?
rn when I migrate to 18/19 this keeps triggering.

**Pull request merged**